Documentation/design-docs/dllmap.md

```diff
@@ -0,0 +1,246 @@
+
+### Entrypoint mapping
+Once hmod gets resolved and returned via `LoadLibraryModule`, an entrypoint must be find:
+
+Currently `NDirectGetEntryPoint` takes two arguments: `pMD` and hmod. When `NDirectGetEntryPoint`  gets called, `pMD `points to a target dll and hmod is correlated with a target dll too.
+The entrypoint name mapping will be done at the beginning of `NDirectGetEntryPoint`. A new method ` IntPtr  GetMappedEntrypoint(sourceEntrypointName)`  will be a callback and will return a mapping for an entrypoint if it exists. `pMD` will get updated to point to a target entrypoint. The rest of `NDirectGetEntryPoint()`  flow will remain the same.
+```c++
+HINSTANCE hmod = LoadLibraryModule( pMD, &errorTracker );
+if ( hmod )
+{
+	LPVOID pvTarget = GetEntrypointViaCallback(pMD, hmod); // this is the only introduced step
+	if (!pvTarget)
+	{
+		LPVOID pvTarget = NDirectGetEntryPoint(pMD, hmod);
+	}
+	…
+}
+```
+
+In consequence, after `LoadLibraryModule()`  and `GetEntrypointViaCallback() / NDirectGetEntryPoint()`  execution, `pMD` will get updated: `pMD->SetNDirectTarget(pvTarget)` with a correctly mapped `pvTarget`.
+
+### Callbacks
+
+As explained above, runtime will rise two dll specific events on each load attempt:
+* 1st callback - when loading a non-system library
+* 2nd callback - when finding an entrypoint
+
+Events will be defined in `AssemblyLoadContext` in `Private.CoreLib`. Default handlers that subscribe dll load events will implement the mono-based dllmap logic. They will take string as argument and return IntPtr of target libraries and entrypoints based on the parsed XML configuration file.
+
+Handlers implementation will stay in `AssemblyLoadContext`. Load library resolver will cache all the dll mapping results that got resolved (as key-value: IntPtr-hmod pairs). Thanks to that, the same library won't get loaded multiple times. User’s code will be able to subscribe to events and implement any loading behavior. That will give a user full flexibility when using dllmap and won’t limit defining the mapping to only xml-based style. Callbacks will be executed for non-system assemblies only. `System.Private.CoreLib` and other system assemblies will not receive callbacks, because that could cause problems like an infinite dll-mapping loop.
+
+### Resolution flow
+
+**User’s code [managed code]**
+
+-   Subscribes to `LoadLibraryModuleViaCallback` and `GetEntrypointViaCallback` events with their default or custom handler
+-   Uses DllImport directive and does the p/invoke
+	 ```c++
+	[DllImport("MyLibrary.dll", EntryPoint="MyFunction")]
+	static  extern  int  MyFunction();
+	…
+	System.Runtime.Loader.AssemblyLoadContext.Default.ResolveNativeDllName += LoadNativeDllViaDllMap;
```

Comment:
These APIs will need to go through https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md before they are committed to master